src/patterns/sns-lambda.test.ts

```diff
@@ -0,0 +1,71 @@
+import "@aws-cdk/assert/jest";
+import { SynthUtils } from "@aws-cdk/assert";
+import { Runtime } from "@aws-cdk/aws-lambda";
+import type { SynthedStack } from "../../test/utils";
+import { simpleGuStackForTesting } from "../../test/utils";
+import type { NoMonitoring } from "../constructs/cloudwatch/no-monitoring";
+import { GuSnsLambda } from "./sns-lambda";
+
+describe("The GuSnsLambda pattern", () => {
+  it("should create the correct resources for a new stack with minimal config", () => {
+    const stack = simpleGuStackForTesting();
+    const noMonitoring: NoMonitoring = { noMonitoring: true };
+    const props = {
+      code: { bucket: "test-dist", key: "lambda.zip" },
+      functionName: "my-lambda-function",
+      handler: "my-lambda/handler",
+      runtime: Runtime.NODEJS_12_X,
+      monitoringConfiguration: noMonitoring,
+    };
```

Comment:
Possibly not for addressing in this PR, but I think these props are too big. I wonder if we could default `code`, `handler` and `runtime` with sensible defaults but still allow one to override them.

This will, for example, make it easier to keep to the latest node runtime environment. Version X of the CDK library will default the runtime to v12. When AWS support a newer runtime, version Y of CDK gets published and teams update their stacks by bumping a version number in package.json and get the updated runtime for free.

Reply:
> I wonder if we could default code, handler and runtime with sensible defaults but still allow one to override them.

I worry that with the current distribution of projects between Scala and TypeScript/JavaScript there is no single sensible default here. Perhaps we can take [RuntimeFamily](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda.RuntimeFamily.html) as a prop, and produce sensible defaults based on that. We already do something similar for memory size [here](https://github.com/guardian/cdk/blob/main/src/constructs/lambda/lambda.ts#L25-L36).

> When AWS support a newer runtime, version Y of CDK gets published and teams update their stacks by bumping a version number in package.json and get the updated runtime for free.

I think this will be great technique for ensuring that new projects always use the latest runtime for their 'family'. For existing projects, I guess that a major Node upgrade may also require code changes / more careful testing, so we can probably do this as long as we treat it as a major version change?

Overall, I think it makes sense to implement this type of abstraction at the `GuLambdaFunction` (construct) level. So I'd be inclined to create a new card and tackle it in a separate PR, if that's OK with you?

Reply:
Another version of this just came to mind: make `GuSnsLambda` abstract and then publish a `GuSnsNodeLambda` and `GuSnsJVMLambda` that extend `GuSnsLambda`. If consumers want a different runtime, they can extend `GuSnsLambda` themselves. Making it similar to [`GuPolicy`, `GuAllowPolicy` and `GuDenyPolicy`](https://github.com/guardian/cdk/blob/404bd66d9c5c06a851b885f22490ecca397d5c18/src/constructs/iam/policies/base-policy.ts#L11-L42).

https://github.com/guardian/cdk/blob/404bd66d9c5c06a851b885f22490ecca397d5c18/src/constructs/iam/policies/base-policy.ts#L11-L42

> So I'd be inclined to create a new card and tackle it in a separate PR, if that's OK with you?

Absolutely! 👍🏽

Reply:
I've created a [card](https://trello.com/c/vrpIdRIX/530-reduce-number-of-required-props-for-gulambdafunction) to remind us about this potential improvement.